pkg/package-server/provider/registry.go

```diff
@@ -131,43 +158,176 @@ func (p *RegistryProvider) syncCatalogSource(obj interface{}) (syncError error)
-		client.conn.ResetConnectBackoff()
+	address := source.Address()
+	logger = logger.WithField("address", address)
 
-		ctx, cancel := context.WithTimeout(context.TODO(), defaultConnectionTimeout)
-		defer cancel()
+	key := resolver.CatalogKey{
+		Namespace: source.GetNamespace(),
+		Name:      source.GetName(),
+	}
+	if sourceMeta := p.sources.GetMeta(key); sourceMeta != nil && sourceMeta.Address == address {
+		// If the address hasn't changed, don't bother creating a new source
+		logger.Debug("catalog address unchanged, skipping source creation")
+		return
+	}
 
-		changed := client.conn.WaitForStateChange(ctx, connectivity.TransientFailure)
-		if !changed {
-			logger.Debugf("grpc connection reset timeout")
-			syncError = fmt.Errorf("grpc connection reset timeout")
-			return
+	logger.Info("connecting to source")
+	if _, syncError = p.sources.Add(key, address); syncError != nil {
+		logger.Warn("failed to create a new source")
+	}
+
+	return
+}
+
+func (p *RegistryProvider) syncSourceState(state registrygrpc.SourceState) {
+	key := state.Key
+	logger := logrus.WithFields(logrus.Fields{
+		"action": "sync source",
+		"source": key,
+		"state":  state.State,
+	})
+	logger.Debug("source state changed")
+
+	timeout, cancel := context.WithTimeout(context.Background(), cacheTimeout)
+	defer cancel()
+
+	var err error
+	switch state.State {
+	case connectivity.Ready:
+		var client *registryClient
+		client, err = p.registryClient(key)
+		if err == nil {
+			err = p.refreshCache(timeout, client)
 		}
+	case connectivity.TransientFailure, connectivity.Shutdown:
+		err = p.gcPackages(key, nil)
+	default:
+		logger.Debug("inert source state, skipping cache update")
+	}
 
-		logger.Info("grpc connection reset")
+	if err != nil {
+		logger.WithError(err).Warn("failed to update cache")
+	}
+}
+
+func (p *RegistryProvider) registryClient(key resolver.CatalogKey) (client *registryClient, err error) {
+	source := p.sources.Get(key)
+	if source == nil {
+		err = fmt.Errorf("missing source for catalog %s", key)
 		return
-	} else if ok {
-		// Address type grpc CatalogSource, drop the connection dial in to the new address
-		client.conn.Close()
 	}
 
-	logger.Info("attempting to add a new grpc connection")
-	conn, err := grpc.Dial(source.Address(), grpc.WithInsecure())
-	if err != nil {
-		logger.WithField("err", err.Error()).Errorf("could not connect to registry service")
-		syncError = err
+	conn := source.Conn
+	if conn == nil {
+		err = fmt.Errorf("missing grpc connection for source %s", key)
 		return
 	}
 
-	p.setClient(newRegistryClient(source, conn), key)
-	logger.Info("new grpc connection added")
+	var catsrc *operatorsv1alpha1.CatalogSource
+	catsrc, err = p.catsrcLister.CatalogSources(key.Namespace).Get(key.Name)
+	if err != nil {
+		return
+	}
 
+	client = newRegistryClient(catsrc, conn)
 	return
 }
 
+func (p *RegistryProvider) refreshCache(ctx context.Context, client *registryClient) error {
```

Comment:
Decided to pass this a `*registryClient` instead of a key to make testing easier, since we have pre-existing tests that use a fake of this type to directly inject registry responses.